src/NuGet.Core/NuGet.Protocol.Core.v3/Utility/DownloadTimeoutStream.cs

```diff
@@ -0,0 +1,104 @@
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Protocol.Core.v3;
+
+namespace NuGet.Protocol
+{
+    public class DownloadTimeoutStream : Stream
+    {
+        private readonly string _downloadName;
+        private readonly Stream _networkStream;
+        private readonly TimeSpan _timeout;
+
+        public DownloadTimeoutStream(string downloadName, Stream networkStream, TimeSpan timeout)
+        {
+            _downloadName = downloadName;
+            _networkStream = networkStream;
+            _timeout = timeout;
+        }
+
+        public override void Flush()
+        {
+            throw new NotSupportedException();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            try
+            {
+                return ReadAsync(buffer, offset, count).Result;
```

Comment:
I believe internally it redirects to overridden method with a token. But I'd rather make it obvious here and would call it directly with CancellationToken.None.